### PR TITLE
feat: support general WASM module in engine

### PR DESCRIPTION
**Pull request by PPpro**
Re: https://github.com/cocos/3d-tasks/issues/6330

### Changelog

* support `pal/wasm` module exporting `isntantiateWASM` interface
* support `sys.hasFeature(sys.Feature.WASM)` query
* support `WASM_SUPPORT_MODE` constant in `internal:constants` module
* optimize wasm module `bullet` `physx` and `webgpu`
* support wasm in Editor and Preview

### Relative PR
- editor support: https://github.com/cocos/cocos-editor/pull/1893
- external wasm lib: https://github.com/cocos/cocos-engine-external/pull/346
- build tool support: https://github.com/cocos/cocos-ccbuild/pull/5
- babel version fix: https://github.com/cocos/creator-programming/pull/36

### Tested Platform
- [x] Web Desktop
- [x] Wechat
- [x] Native
- [ ] Open Harmony (cannot run asmjs successfully for now, mark it as TODO https://github.com/cocos/3d-tasks/issues/16338)

### Tested Module
- [x] physx (we only use asmjs for the current implementation)
- [x] bullet
- [x] webgpu (tested wasm building)

### WASM docs

https://github.com/cocos/3d-tasks/blob/master/design-docs/scripting/engine-wasm-module.md

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: pal/system-info/minigame/system-info.ts [hunk 1]
```diff
@@ -145,6 +145,7 @@ class SystemInfo extends EventTarget {
             [Feature.EVENT_HANDLE]: this.isXR,
             [Feature.EVENT_HMD]: this.isXR,
             [Feature.EVENT_HANDHELD]: false,
+            [Feature.WASM]: WECHAT,
```
Comment (minggo) [inline pal/system-info/minigame/system-info.ts:148]: WECHAT is bool value?

File: pal/system-info/web/system-info.ts [hunk 2]
```diff
@@ -198,6 +198,26 @@ class SystemInfo extends EventTarget {
         const supportMouse = document.documentElement.onmouseup !== undefined || EDITOR;
         // NOTE: xr is not totally supported on web
         const supportXR = typeof (navigator as any).xr !== 'undefined';
+        // refer https://stackoverflow.com/questions/47879864/how-can-i-check-if-a-browser-supports-webassembly
+        const supportWasm = (() => {
+            // iOS 15.4 has some wasm memory issue, can not use wasm for bullet
+            const isiOS15_4 = (this.os === OS.IOS || this.os === OS.OSX) && /(OS 15_4)|(Version\/15.4)/.test(window.navigator.userAgent);
```
Comment (minggo) [inline pal/system-info/web/system-info.ts]: But why check OSX too?

File: pal/system-info/web/system-info.ts [hunk 3]
```diff
@@ -198,6 +198,26 @@ class SystemInfo extends EventTarget {
...
+            if (isiOS15_4) {
+                return false;
+            }
+            try {
+                if (typeof WebAssembly === 'object'
+                    && typeof WebAssembly.instantiate === 'function') {
+                    const module = new WebAssembly.Module(new Uint8Array([0x0, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00]));
+                    if (module instanceof WebAssembly.Module) {
```
Comment (minggo) [inline pal/system-info/web/system-info.ts:212]: What does it mean? It seems the module includes magic number. Are they magic number for wasm file header?

File: pal/wasm/wasm-web.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ THE SOFTWARE.
+*/
+
+import { EDITOR, PREVIEW } from "internal:constants";
+
+export function instantiateWasm (wasmUrl: string, importObject: WebAssembly.Imports): Promise<any> {
+    // NOTE: when it's in EDITOR or PREVIEW, wasmUrl is an absolute file path of wasm.
+    if (EDITOR) {
+        // IDEA: it's better we implement another PAL for nodejs platform.
+        const fs = require('fs');
```
Comment (minggo) [inline pal/wasm/wasm-web.ts:34]: Can use nodejs in engine codes?

File: pal/system-info/minigame/system-info.ts (hunk 1 above)
Comment (PPpro) [reply pal/system-info/minigame/system-info.ts:148]: yes

File: pal/system-info/web/system-info.ts (hunk 2 above)
Comment (PPpro) [reply pal/system-info/web/system-info.ts]: it's a history bug fix, it seems that on safari, there is some memory issue

File: pal/system-info/web/system-info.ts (hunk 3 above)
Comment (PPpro) [reply pal/system-info/web/system-info.ts:212]: yes, it's a simple test to detect whether the browser supports WASM

they are just some test data

File: pal/wasm/wasm-web.ts (hunk 4 above)
Comment (PPpro) [reply pal/wasm/wasm-web.ts:34]: here we only use on Editor environment

File: pal/system-info/web/system-info.ts (hunk 2 above)
Comment (minggo) [reply pal/system-info/web/system-info.ts]: But the variable name is `isiOS15_4 `.

Comment (PPpro) [reply pal/system-info/web/system-info.ts]: yes, it means on safari iOS15.4

Comment (PPpro) [reply pal/system-info/web/system-info.ts]: updated here https://github.com/cocos/cocos-engine/pull/14974/commits/0e1101f954a3a3bd23a1ba0f470516c6613bb4b0

File: pal/system-info/web/system-info.ts [hunk 5]
```diff
@@ -200,9 +200,9 @@ class SystemInfo extends EventTarget {
         const supportXR = typeof (navigator as any).xr !== 'undefined';
         // refer https://stackoverflow.com/questions/47879864/how-can-i-check-if-a-browser-supports-webassembly
         const supportWasm = (() => {
-            // iOS 15.4 has some wasm memory issue, can not use wasm for bullet
-            const isiOS15_4 = (this.os === OS.IOS || this.os === OS.OSX) && /(OS 15_4)|(Version\/15.4)/.test(window.navigator.userAgent);
-            if (isiOS15_4) {
+            // safari iOS 15.4 has some wasm memory issue, can not use wasm for bullet
```
Comment (minggo) [inline pal/system-info/web/system-info.ts]: I think should modify the comment too? As not only has issue on iOS.

Comment (PPpro) [reply pal/system-info/web/system-info.ts]: udpated https://github.com/cocos/cocos-engine/pull/14974/commits/de35e29db68b7871de0d9e093f3b58e29f41fac8

**Review by minggo — APPROVED** [review]

**Review by hana-alice — APPROVED** [review]

**Review by lealzhan — APPROVED** [review]

**Review by yiwenxue — APPROVED** [review]

File: cocos/misc/webassembly-support.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import {} from 'internal:constants';
+
+/**
+ * An enum type for WASM_SUPPORT_MODE constant, making the support mode more readable.
+ */
+export const enum WebAssemblySupportMode {
+    NONE = 0,
+    SUPPORT = 1,
+    MAYBE_SUPPORT = 2,
+}
```
Comment (PPpro) [inline cocos/misc/webassembly-support.ts]:
```suggestion
/**
 * An enum type for WASM_SUPPORT_MODE constant, making the support mode more readable.
 */
export const enum WebAssemblySupportMode {
    NONE = 0,
    SUPPORT = 1,
    MAYBE_SUPPORT = 2,
}
```

File: cocos/physics/bullet/instantiated.ts [hunk 7]
```diff
@@ -70,74 +66,46 @@ export const bt: instanceExt = {} as any;
 globalThis.Bullet = bt;
 bt.BODY_CACHE_NAME = 'body';
 
+function initWasm (wasmUrl: string) {
+    console.info('[Physics][Bullet]: Using wasm Bullet libs.');
```
Comment (shrinktofit) [inline cocos/physics/bullet/instantiated.ts]: Use engine `debug()` instead

File: cocos/physics/bullet/instantiated.ts [hunk 8]
```diff
@@ -70,74 +66,46 @@ export const bt: instanceExt = {} as any;
...
+    };
+
+    return instantiateWasm(wasmUrl, importObject).then((results) => {
+        const btInstance = results.instance.exports as Bullet.instance;
+        Object.assign(bt, btInstance);
+    });
+}
+
+function initAsm (resolve) {
+    console.info('[Physics][Bullet]: Using asmjs Bullet libs.');
```
Comment (shrinktofit) [inline cocos/physics/bullet/instantiated.ts]: Same

File: cocos/physics/physx/physx-adapter.ts [hunk 9]
```diff
@@ -61,23 +64,58 @@ export function InitPhysXLibs () {
...
 
-function initPhysXWithJsModule () {
-    // If external PHYSX not given, then try to use internal PhysX libs.
-    globalThis.PhysX = globalThis.PHYSX ? globalThis.PHYSX : PhysX;
+function initASM () {
+    globalThis.PhysX = globalThis.PHYSX ? globalThis.PHYSX : asmFactory;
     if (globalThis.PhysX != null) {
         return globalThis.PhysX().then((Instance: any) => {
-            if (!EDITOR && !TEST) console.info('[PHYSICS]:', `${USE_EXTERNAL_PHYSX ? 'External' : 'Internal'} PhysX libs loaded.`);
+            if (!EDITOR && !TEST) console.info('[PHYSICS]:', `${USE_EXTERNAL_PHYSX ? 'External' : 'Internal'} PhysX asm libs loaded.`);
```
Comment (shrinktofit) [inline cocos/physics/physx/physx-adapter.ts]: Same

**Review by shrinktofit — APPROVED** [review]

**Pull request merged**